### PR TITLE
[AIRFLOW-5660] Attempt to find the task in DB from Kubernetes pod labels

### PR DESCRIPTION
**Pull request by adityav**
…ing every task

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-XXX
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

1. Updates function to first do a task_id/dag_id search of the DB to significantly reduce search time
2. In the `_make_safe_label_value` function we can add a warning if a task_id or dag_id will require hashing (which will slow down processing
3. IF there is no database match we do a full scan.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: airflow/executors/kubernetes_executor.py [hunk 1]
```diff
@@ -545,6 +545,26 @@ def _labels_to_key(self, labels):
...
+            )
+            if task:
+                self.log.info(
+                    'Found matching task %s-%s (%s) with current state of %s',
+                    task.dag_id, task.task_id, task.execution_date, task.state
+                )
+                return (dag_id, task_id, ex_time, try_num)
+            else:
+                self.log.warning(
+                    "Unable to find Task in db directly for %s-%s (%s). Scan all tasks",
```
Comment (dimberman) [inline airflow/executors/kubernetes_executor.py]: can you emphasize that this can cause significant slowdowns?

Comment (dimberman) [reply airflow/executors/kubernetes_executor.py]: And add a link to k8s labeling guidelines?  https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set

**Review by dimberman — CHANGES_REQUESTED** [review]

File: airflow/executors/kubernetes_executor.py (hunk 1 above)
Comment (adityav) [reply airflow/executors/kubernetes_executor.py]: sure. I will rework the code a bit to check for this explicitly.

**Review by dimberman — APPROVED** [review]

**Comment by dimberman** [comment]
LGTM. Rerunning tests.

**Comment by adityav** [comment]
Looks like there is a problem with CI? I have no idea why the build failed.

File: airflow/executors/kubernetes_executor.py [hunk 2]
```diff
@@ -545,6 +545,27 @@ def _labels_to_key(self, labels):
             return None
 
         with create_session() as session:
+            task = (
+                session
+                .query(TaskInstance)
+                .filter_by(task_id=task_id, dag_id=dag_id, execution_date=ex_time)
+                .first()
```
Comment (ashb) [inline airflow/executors/kubernetes_executor.py]:
```suggestion
                .one_or_none()
```

**Comment by ashb** [comment]
Have you run this in a Kube cluster? I have a feeling that _every_ task will hit the bad path because of the characters in the execution date

**Comment by adityav** [comment]
> Have you run this in a Kube cluster? I have a feeling that _every_ task will hit the bad path because of the characters in the execution date

We are running airflow in EKS with this patch applied and it works. We can finally scale to 100,000+ tasks with this. Previously it would choke with 5k-10k tasks.
Only dag_id / task_id are being stored in the labels. The execution date isn't being stored there so it shouldn't be a problem. 

Ideally I would prefer to eliminate the bad path altogether. Currently, it requires the dag writer to write good dag id / task id which isn't a good design. I can only think of 2 solns:

1. Use task_id / dag_id stored in env variables. Values stored in env variables don't have any label specific restrictions. However, I am not familiar with kube api to know if this is a possible soln or not.
2. have a mapping table of (exec_date, dag_id, task_id, safe_dag_id, safe_task_id) in airflow metadb.

**Comment by ashb** [comment]
> Only dag_id / task_id are being stored in the labels. The execution date isn't being stored there so it shouldn't be a problem.

https://github.com/apache/airflow/blob/5a0ddc576a45e7c2e573eb468f4f133bd56cc03b/airflow/executors/kubernetes_executor.py#L539 begs to differ. But `_label_safe_datestring_to_datetime` and it's reverse means it's probably okay:)

Final question: which log file does this warning show up in?

**Comment by adityav** [comment]
> > Only dag_id / task_id are being stored in the labels. The execution date isn't being stored there so it shouldn't be a problem.
> 
> https://github.com/apache/airflow/blob/5a0ddc576a45e7c2e573eb468f4f133bd56cc03b/airflow/executors/kubernetes_executor.py#L539
> begs to differ. But `_label_safe_datestring_to_datetime` and it's reverse means it's probably okay:)

My bad, didn't see that.

> Final question: which log file does this warning show up in?

scheduler logs. They are in the console output of the scheduler pod.

File: airflow/executors/kubernetes_executor.py [hunk 3]
```diff
@@ -545,6 +545,27 @@ def _labels_to_key(self, labels):
...
+                    task.dag_id, task.task_id, task.execution_date, task.state
+                )
+                return (dag_id, task_id, ex_time, try_num)
+            else:
+                self.log.warning(
+                    'task_id/dag_id are not safe to use as Kubernetes labels. This can cause '
+                    'severe performance regressions. Please see '
+                    'https://kubernetes.io/docs/concepts/overview/working-with-objects'
+                    '/labels/#syntax-and-character-set'
+                    'Given dag_id: %s, task_id: %s', task_id, dag_id
```
Comment (ashb) [inline airflow/executors/kubernetes_executor.py:566]: Missing a space here at the start - should probably also wrap the URL in `<>`

**Review by ashb — CHANGES_REQUESTED** [review]
(sorry for leaving this languishing for so long)

**Review by ashb — APPROVED** [review]

**Pull request merged**